### ISSUE TITLE
Upgrade to ECMAScript 6

### ISSUE THREAD

**Issue by thallgren** [robertkrimen/otto#313]
[ECMAScript 6](http://es6-features.org/) contains a lot of improvements and additions to JavaScript. It would be great if those could be implemented in otto.

**Comment by q2dg** [comment]
Well, it seems this project is pretty abandoned. This is very sad, because it's used in great software like Bettercap

**Comment by deoxxa** [comment]
Nope, not abandoned. Everyone involved here has a full time job though, so nobody's really got the time to invest in new features. If you'd like to implement ES6 (lol that's a gigantic job), feel free to ask questions about the implementation.

**Comment by twitchyliquid64** [comment]
Can you comment a little on what you think the big challenges would be for implementing this?

Obviously the es6 spec has a lot of new features, but most of them seem like relatively straightforward additions to the AST, and then a medium amount of work plugging it into the runtime / `_object`.

I'm down to give this a go sometime in the near future.

**Comment by Asday** [comment]
Is there actually a reason to implement ES6?  From what I understand, everything in ES6 (including the non-standard stuff) can be transpiled down to ES5 using babel (or webpack?  I get confused), to be run under otto.

**Comment by twitchyliquid64** [comment]
Aside from execution, otto exposes a parser and other primitives that would be useful for ES6. Not to mention supporting ES6 would mean that people can run ES6 natively, which would be a nice feature.

Regardless, I dont agree with that reasoning, as inputs can be transpiled down for any turing-complete interpreter.

In any case, I think its safe to call this repo abandoned (multiple emails to maintainers about this have gone unreplied, no further action/update since the 'nope not abandoned' comment).

**Comment by Asday** [comment]
>would be useful for ES6

How?

>which would be a nice feature

Reminder that every maintainer of this project does so for free in their spare time for fun.  They're not here to provide "nice features" that aren't even needed.

>inputs can be transpiled down for any turing-complete interpreter.

That's pedantry and you're very aware of it.  ES6 is already regularly transpiled to ES5 to run under old browsers.  The tooling is in-place and very mature.

>I think its safe to call this repo abandoned

Our good friend @deoxxa commented that it indeed _isn't_ abandoned.  What exactly do you want from them regarding this ticket?  They said they'd help out with implementation details, but you don't appear to want to do the work.

**Comment by thallgren** [comment]
For the record, the reason I created this ticket was that the project parser produces an AST that can be used for purposes other than just evaluating, such as transforming into ASTs tailored for other types of evaluators or compilers. I'm not at all interested in code that has been transpiled to ES5.

**Comment by twitchyliquid64** [comment]
I apologize if I came across as flippant, that certainty wasn't my intention.

> > would be useful for ES6
> 
> How?

EG: Code analysis needs to have an accurate (ie: not lowered) representation of the code in question. If your lexer/parser/AST cannot represent ES6 features, then you cant use such tools with ES6 code. A specific example would be a linter for code style.

> That's pedantry and you're very aware of it.

Hardly. Consider interoperability between go code and ES6-only features like modules. In an ES5 environment (and indeed Otto in its current state), theres `vm.Run()`, but for ES6 would be exposing much richer, module-aware APIs. What about proxys? I'm not sure if transpilers work there (could be wrong thou). It would be great for otto to allow native objects to have callbacks for this, which would be entirely impossible with an emulated representation.

I'm no expert on ES6 transpilers, but i would expect some translated features to have non-neglible runtime performance hit - which is another benefit of native support.

I acknowledge some number of users could get by with a transpiler, but there are edge cases and downsides here, and my position is that these edge cases are sufficient justification for a willing party to implement support.

> They're not here to provide "nice features" that aren't even needed.

> They said they'd help out with implementation details, but you don't appear to want to do the work.

These comments seem to suggest that I expect the maintainers to do anything. I do not. Maintainers of OSS projects are under no obligation to spend their free time on free software. They can take the project in any direction or no direction they feel fit.

> What exactly do you want from them

I've expressed my interest on implementing ES6 and am willing to do the work. I followed up both in this thread and via email WRT PR/reviews and high-level challenges, and received no response.

**Comment by twitchyliquid64** [comment]
> such as transforming into ASTs tailored for other types of evaluators or compilers

Out of curiosity, what were you targeting?

**Comment by thallgren** [comment]
@twitchyliquid64 I worked in a project where we wanted definitions to be polyglot. The intention was to use different language parsers and then transform their ASTs into a common AST that could be evaluated by our own evaluator. Eventually we decided to take another path and use the native environment for each targeted language.

**Comment by bysir-zl** [comment]
Thx all contributors of otto. You've done a good job so far .
